caas/kubernetes/provider/k8s.go

```diff
@@ -877,15 +904,54 @@ func processConstraints(pod *core.PodSpec, appName string, cons constraints.Valu
 	return nil
 }
 
-// ApplyRawK8sSpec applies raw k8s spec to the k8s cluster.
-func (k *kubernetesClient) ApplyRawK8sSpec(specStr string) error {
-	spec, err := k8sspecs.ParseRawK8sSpec(specStr)
-	if err != nil {
+const applyRawSpecTimeoutSeconds = 20
+
+func (k *kubernetesClient) applyRawK8sSpec(
+	appName string,
+	statusCallback caas.StatusCallbackFunc,
+	params *caas.ServiceParams,
+	numUnits int,
+	config application.ConfigAttributes,
+) (err error) {
+
+	if params == nil || len(params.RawK8sSpec) == 0 {
+		return errors.Errorf("missing raw pod spec")
+	}
+
+	if params.Deployment.DeploymentType == "" {
+		params.Deployment.DeploymentType = caas.DeploymentStateless
+		if len(params.Filesystems) > 0 {
+			params.Deployment.DeploymentType = caas.DeploymentStateful
+		}
+	}
+
+	// TODO(caas): support Constraints, FileSystems, Devices, InitContainer for actions, etc.
+	if err := params.Deployment.DeploymentType.Validate(); err != nil {
 		return errors.Trace(err)
 	}
-	logger.Debugf("specStr -> %s", specStr)
-	logger.Debugf("spec -> %s", pretty.Sprint(spec))
-	return errors.NotImplementedf("raw k8s spec")
+
+	logger.Debugf("creating/updating application %s", appName)
+	deploymentName := k.deploymentName(appName, false)
+
+	if numUnits < 0 {
```

Comment:
numUnits is checked here but not used below when applying the raw spec.
I think it should set the replicas count

Reply:
As we discussed and the TODO 2 mentioned, the `.Spec` schema is not parsed in this PR yet which causes a few key features that are missing: Scale, Constraints, FileSystems, Devices, InitContainer for actions etc.